src/main/scala/scorex/core/NodeViewHolder.scala

```diff
@@ -1,393 +1,306 @@
+import scorex.core.NodeViewComponentOperation.GetReader
+import scorex.core.consensus.{HistoryReader, SyncInfo}
 import scorex.core.network.NodeViewSynchronizer.ReceivableMessages.NodeViewHolderEvent
 import scorex.core.settings.ScorexSettings
 import scorex.core.transaction._
-import scorex.core.transaction.state.{MinimalState, TransactionValidation}
-import scorex.core.transaction.wallet.Vault
-import scorex.core.utils.ScorexEncoding
-import scorex.util.ScorexLogging
-import scala.annotation.tailrec
-import scala.util.{Failure, Success, Try}
+import scorex.core.transaction.state.{MinimalState, StateHistoryActor, StateReader}
+import scorex.core.transaction.wallet.VaultReader
+import scorex.util.{ScorexEncoding, ScorexLogging}
 
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success}
 
-/**
-  * Composite local view of the node
+
+/** Composite local view of the node.
   *
   * Contains instances for History, MinimalState, Vault, MemoryPool.
   * The instances are read-only for external world.
   * Updates of the composite view(the instances are to be performed atomically.
   *
-  * @tparam TX
-  * @tparam PMOD
+  * The main data structure a node software is taking care about, a node view consists
+  * of four elements to be updated atomically: history (log of persistent modifiers),
+  * state (result of log's modifiers application to pre-historical(genesis) state,
+  * user-specific information stored in vault (it could be e.g. a wallet), and a memory pool.
   */
 trait NodeViewHolder[TX <: Transaction, PMOD <: PersistentNodeViewModifier]
   extends Actor with ScorexLogging with ScorexEncoding {
 
   import NodeViewHolder.ReceivableMessages._
   import NodeViewHolder._
   import scorex.core.network.NodeViewSynchronizer.ReceivableMessages._
+  import scorex.core.transaction.MempoolOperation._
+  import scorex.core.transaction.state.StateOperation._
 
   type SI <: SyncInfo
-  type HIS <: History[PMOD, SI, HIS]
-  type MS <: MinimalState[PMOD, MS]
-  type VL <: Vault[TX, PMOD, VL]
-  type MP <: MemoryPool[TX, MP]
+  type History <: consensus.History[PMOD, SI, History]
+  type State <: MinimalState[PMOD, State]
+  type Vault <: transaction.wallet.Vault[TX, PMOD, Vault]
+  type MPool <: MempoolReader[TX]
 
-  type NodeView = (HIS, MS, VL, MP)
+  /** State actor used for asynchronous transaction validation
+    */
+  protected val stateActor: ActorRef = createStateActor()
 
-  case class UpdateInformation(history: HIS,
-                               state: MS,
-                               failedMod: Option[PMOD],
-                               alternativeProgressInfo: Option[ProgressInfo[PMOD]],
-                               suffix: IndexedSeq[PMOD])
+  /** Underlying memory pool actor
+    */
+  protected val memoryPoolActor: ActorRef = createMemoryPoolActor()
+
+  protected var vault: Vault = restoreVault().getOrElse(genesisVault())
```

Comment:
Why isn't `vault` implemented as an actor?

Reply:
Yes, this definitely should be implemented as an actor, I just did not manage to do it in time. Please fix this